### PR TITLE
Improve compiler argument handling with a new class CompilerArgs derived from list()

### PR DESCRIPTION
**Pull request by nirbheek**
The purpose of this class is to make it possible to sanely generate compiler command-lines by ensuring that new arguments appended or added to a list of arguments properly override previous arguments. For instance:

```
>>> a = CompilerArgs(['-Lfoo', '-DBAR'])
>>> a += ['-Lgah', '-DTAZ']
>>> a
['-Lgah', '-Lfoo', '-DBAR', '-DTAZ']
```

Arguments will be de-duped if it is safe to do so. Currently, this is only done for `-I`, `-D` and `-L` arguments (previous occurances are removed when a new one is added) and arguments that once added cannot be overriden such as `-pipe` are removed completely.

The order in which compiler arguments are added for single-compiles has been changed significantly and a test has been added for the same.

The order in which compiler arguments are added for linking has been tweaked a little bit, but more needs to be done. See the commit message for details.

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -57,6 +59,62 @@ def test_version_number(self):
...
+        # Test that empty initialization works
+        a = cargsfunc(c)
+        self.assertEqual(a, [])
+        # Test that list initialization works
+        a = cargsfunc(['-I.', '-I..'], c)
+        self.assertEqual(a, ['-I.', '-I..'])
+        # Test that there is no de-dup on initialization
+        self.assertEqual(cargsfunc(['-I.', '-I.'], c), ['-I.', '-I.'])
+
+        ## Test that appending works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: run_unittests.py [hunk 2]
```diff
@@ -57,6 +59,62 @@ def test_version_number(self):
...
+        a.append('-I..')
+        self.assertEqual(a, ['-I..', '-I.'])
+        a.append('-O3')
+        self.assertEqual(a, ['-I..', '-I.', '-O3'])
+
+        ## Test that in-place addition works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: run_unittests.py [hunk 3]
```diff
@@ -57,6 +59,62 @@ def test_version_number(self):
...
+        # Test that removal works
+        a.remove('-O2')
+        self.assertEqual(a, ['-I..', '-I.', '-O3', '-O2'])
+        # Test that de-dup happens on addition
+        a += ['-Ifoo', '-Ifoo']
+        self.assertEqual(a, ['-Ifoo', '-I..', '-I.', '-O3', '-O2'])
+
+        # .extend() is just +=, so we don't test it
+
+        ## Test that addition works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: run_unittests.py [hunk 4]
```diff
@@ -57,6 +59,62 @@ def test_version_number(self):
...
+        a = a + ['-Ifoo']
+        self.assertEqual(a, ['-Ifoo', '-I..', '-I.', '-O3', '-O2'])
+        # Test that adding a list with one arg new and one old works
+        a = a + ['-Ifoo', '-Ibaz']
+        self.assertEqual(a, ['-Ifoo', '-Ibaz', '-I..', '-I.', '-O3', '-O2'])
+        # Test that adding args that must be prepended and appended works
+        a = a + ['-Ibar', '-Wall']
+        self.assertEqual(a, ['-Ibar', '-Ifoo', '-Ibaz', '-I..', '-I.', '-O3', '-O2', '-Wall'])
+
+        ## Test that reflected addition works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: mesonbuild/backend/ninjabackend.py [hunk 5]
```diff
@@ -2046,17 +2070,75 @@ def generate_link(self, target, outfile, outname, obj_list, linker, extra_args=[
...
+        if not isinstance(target, build.StaticLibrary):
+            # Add link args added using add_project_link_arguments()
+            commands += self.build.get_project_link_args(linker, target.subproject)
+            # Add link args added using add_global_link_arguments()
+            # These override per-project link arguments
+            commands += self.build.get_global_link_args(linker)
+            # Link args added from the env: LDFLAGS. We want these to
+            # override all the defaults but not the per-target link args.
+            commands += self.environment.coredata.external_link_args[linker.get_language()]
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [W293] blank line contains whitespace

**Comment by ignatenkobrain** [comment]
vs backend is broken:
```python
Traceback (most recent call last):
  File "C:\projects\meson\mesonbuild\mesonmain.py", line 295, in run
    app.generate()
  File "C:\projects\meson\mesonbuild\mesonmain.py", line 179, in generate
    g.generate(intr)
  File "C:\projects\meson\mesonbuild\backend\vs2010backend.py", line 164, in generate
    projlist = self.generate_projects()
  File "C:\projects\meson\mesonbuild\backend\vs2010backend.py", line 292, in generate_projects
    self.gen_vcxproj(target, projfile, uuid)
  File "C:\projects\meson\mesonbuild\backend\vs2010backend.py", line 690, in gen_vcxproj
    file_args[l] += compiler.unix_compile_flags_to_native(args)
AttributeError: 'VisualStudioCPPCompiler' object has no attribute 'unix_compile_flags_to_native'
```

**Comment by nirbheek** [comment]
Still need to fix the include order for the VS backends, which is why test 136 is failing. It's on my TODO list.

**Comment by jpakkane** [comment]
Would it be useful to print a warning if there are two conflicting declarations for a single token? Something like

    x = ['-DFOO=BAR']
    x += ['-FOO=BAZ']

**Comment by nirbheek** [comment]
Good question... maybe? However, I know of use-cases where defines are set project-wide and then unset per-target. 

For instance, you need to define `WINVER` and/or `_WIN32_WINNT` to values corresponding to old Windows versions before including the Windows SDK to get access to header prototypes for deprecated API, and to values for newer Windows versions to get access to new API, so you often have to set them per-plugin in places where you use deprecated API.

You definitely want to set them if you use the Windows SDK or you will always target only the Windows version of that SDK (which is usually the latest) and you almost never want that, so people usually have a project-wide "default" that is overriden wherever you use deprecated API.

Maybe we can print that warning only at the highest warning level!

**Comment by jpakkane** [comment]
:crossed_swords: conflicts

**Comment by nirbheek** [comment]
Will fix it, but not ready for merging yet!

File: run_unittests.py [hunk 6]
```diff
@@ -13,12 +13,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+<<<<<<< dbcbf19ecea9d07d264dbbc1cd87ab22393be8a7
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E227] missing whitespace around bitwise or shift operator
* [E901] SyntaxError: invalid syntax
* [E225] missing whitespace around operator

File: run_unittests.py [hunk 7]
```diff
@@ -13,12 +13,17 @@
...
 import stat
+=======
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E225] missing whitespace around operator

File: run_unittests.py [hunk 8]
```diff
@@ -13,12 +13,17 @@
...
+import shlex
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E402] module level import not at top of file

File: run_unittests.py [hunk 9]
```diff
@@ -13,12 +13,17 @@
...
+>>>>>>> Use CompilerArgs for generation of compile commands
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E225] missing whitespace around operator

File: run_unittests.py [hunk 10]
```diff
@@ -13,12 +13,17 @@
...
 import unittest, os, sys, shutil, time
 import subprocess
 import re, json
 import tempfile
 from glob import glob
+import mesonbuild.compilers
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E402] module level import not at top of file

File: mesonbuild/backend/vs2010backend.py [hunk 11]
```diff
@@ -668,90 +670,132 @@ def gen_vcxproj(self, target, ofname, guid):
...
-        for l, comp in target.compilers.items():
-            if l in file_args:
-                file_args[l] += comp.get_option_compile_args(self.environment.coredata.compiler_options)
+                file_args[l] += args
+        # The highest priority includes. In order of directory search:
+        # target private dir, target build dir, generated sources include dirs,
+        # target source dir
+        for args in file_args.values():
+            t_inc_dirs = ['.', self.relpath(self.get_target_private_dir(target),
+                                                 self.get_target_dir(target))] + \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/vs2010backend.py [hunk 12]
```diff
@@ -668,90 +670,132 @@ def gen_vcxproj(self, target, ofname, guid):
...
+                          generated_files_include_dirs + [proj_to_src_dir]
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E126] continuation line over-indented for hanging indent

File: run_unittests.py [hunk 13]
```diff
@@ -95,6 +97,62 @@ def test_mode_symbolic_to_bits(self):
...
+        # Test that empty initialization works
+        a = cargsfunc(c)
+        self.assertEqual(a, [])
+        # Test that list initialization works
+        a = cargsfunc(['-I.', '-I..'], c)
+        self.assertEqual(a, ['-I.', '-I..'])
+        # Test that there is no de-dup on initialization
+        self.assertEqual(cargsfunc(['-I.', '-I.'], c), ['-I.', '-I.'])
+
+        ## Test that appending works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: run_unittests.py [hunk 14]
```diff
@@ -95,6 +97,62 @@ def test_mode_symbolic_to_bits(self):
...
+        a.append('-I..')
+        self.assertEqual(a, ['-I..', '-I.'])
+        a.append('-O3')
+        self.assertEqual(a, ['-I..', '-I.', '-O3'])
+
+        ## Test that in-place addition works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: run_unittests.py [hunk 15]
```diff
@@ -95,6 +97,62 @@ def test_mode_symbolic_to_bits(self):
...
+        # Test that removal works
+        a.remove('-O2')
+        self.assertEqual(a, ['-I..', '-I.', '-O3', '-O2'])
+        # Test that de-dup happens on addition
+        a += ['-Ifoo', '-Ifoo']
+        self.assertEqual(a, ['-Ifoo', '-I..', '-I.', '-O3', '-O2'])
+
+        # .extend() is just +=, so we don't test it
+
+        ## Test that addition works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

File: run_unittests.py [hunk 16]
```diff
@@ -95,6 +97,62 @@ def test_mode_symbolic_to_bits(self):
...
+        a = a + ['-Ifoo']
+        self.assertEqual(a, ['-Ifoo', '-I..', '-I.', '-O3', '-O2'])
+        # Test that adding a list with one arg new and one old works
+        a = a + ['-Ifoo', '-Ibaz']
+        self.assertEqual(a, ['-Ifoo', '-Ibaz', '-I..', '-I.', '-O3', '-O2'])
+        # Test that adding args that must be prepended and appended works
+        a = a + ['-Ibar', '-Wall']
+        self.assertEqual(a, ['-Ibar', '-Ifoo', '-Ibaz', '-I..', '-I.', '-O3', '-O2', '-Wall'])
+
+        ## Test that reflected addition works
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E266] too many leading '#' for block comment

**Comment by nirbheek** [comment]
I've disabled E266 and E402 in this PR.

**Comment by nirbheek** [comment]
PR is ready for merging!

**Pull request merged**